text/0000-let-else.md

```diff
@@ -0,0 +1,652 @@
+
+This pattern serves no practical purpose to a computer, but it is helpful for humans interacting with the code.
+Returning early helps improve code clarity in two ways:
+- Ensuring the returned result in near the conditional, visually, as the following logic may be lengthy.
+- Reduces rightward shift, as the error return is now in the block, rather than the following logic.
+
+This RFC proposes _(Rust provides)_ an extension to `let` assignment statements to help with this pattern, an `else { }` which can follow a pattern match
+as a `let` assigning statement:
+
+```rust
+let Some(a) = an_option else {
+    // Called if `an_option` is not `Option::Some(T)`.
+    // This block must diverge (stop executing the existing context to the parent block or function).
+    return;
+};
+
+// `a` is now in scope and is the type which the `Option` contained.
+```
+
+This is a counterpart to `if let` expressions, and the pattern matching works identically, except that the value from the pattern match
+is assigned to the surrounding scope rather than the block's scope.
+
+# Reference-level explanations
+[reference-level-explanation]: #reference-level-explanation
+
+let-else is syntactical sugar for `match` where the non-matched case diverges.
+```rust
+let pattern = expr else {
+    /* diverging expr */
+};
+```
+desugars to
+```rust
+let (each, binding) = match expr { 
+    pattern => (each, binding),
+    _ => { 
+        /* diverging expr */
+    }
+};
+```
+
+Most expressions may be put into the expression position with two restrictions:
+1. May not include a block outside of parenthesis.
+    - Must be an [`ExpressionWithoutBlock`][expressions].
+    - [`GroupedExpression`][grouped-expr]-s ending with a `}` are additionally not allowed and must be put in parenthesis.
+2. May not be just a lazy boolean expression (`&&` or `||`). (Must not be a [`LazyBooleanExpression`][lazy-boolean-operators].)
+
+While allowing e.g. `if {} else {}` directly in the expression position is technically feasible this RFC proposes it be
+disallowed for programmer clarity so as to avoid `... else {} else {}` situations as discussed in the [drawbacks][] section.
+Boolean matches are not useful with let-else and so lazy boolean expressions are disallowed for reasons noted in [future-possibilities][].
+These types of expressions can still be used when combined in a less ambiguous manner with parenthesis, thus forming a [`GroupedExpression`][grouped-expr],
+which is allowed under the two expression restrictions.
+
+Any refutable pattern that could be put into if-let's pattern position can be put into let-else's pattern position.
+
+If the pattern is irrefutable, rustc will emit the `irrefutable_let_patterns` warning lint, as it does with an irrefutable pattern in an `if let`.
+
+The `else` block must _diverge_, meaning the `else` block must return the [never type (`!`)][never-type]).
+This could be a keyword which diverges (returns `!`), or a panic.
```

Comment:
```suggestion
This could be a keyword which diverges (returns `!`), such as `return`, `break`, `continue` or `loop { ... }`, a diverging function like `std::process::abort`, or a panic.
```

Reply:
`std::process::abort` or `std::process::exit`